src/Adapter/Category/CategoryDataProvider.php

```diff
@@ -167,23 +171,37 @@ public function getCategoriesWithBreadCrumb()
     }
 
     /**
-     * Returns a simple breacrumb from a categoryId, the delimiter can be choosen
-     * @param $categoryId
+     * Construct the breadcrumb using the already constructed list of all categories
+     *
+     * @param int    $categoryId
      * @param string $delimiter
+     *
      * @return string
      */
-    public function getBreadCrumb($categoryId, $delimiter = " > ")
+    public function getBreadCrumb($categoryId, $delimiter = ' > ')
     {
-        $currentCategory = new Category($categoryId);
-        $categories = $currentCategory->getParentsCategories();
+        $categories = $this->getParentNamesFromList($categoryId);
         $categories = array_reverse($categories, true);
-        $breadCrumb = '';
 
-        foreach ($categories as $category) {
-            $breadCrumb .= ' > '.$category['name'];
+        return implode($delimiter, $categories);
+    }
+
+    /**
+     * @param int $categoryId
+     *
+     * @return array
+     */
+    public function getParentNamesFromList($categoryId)
+    {
+        $categories = [];
+
+        while (isset($this->categoryList[$categoryId])) {
```

Comment:
I wonder how you get out of the `while` statement? I don't see any `unset` or `= null` in this method.

Reply:
the $categoryId is updated at each loop to get the parent id value :)